### ISSUE TITLE
HTTP Client does not receive body and gets stuck or silently dies

### ISSUE THREAD

**Issue by LukasForst** [ktorio/ktor#1813]
**Ktor Version and Engine Used (client or server and name)**
1.3.2, client: apache (but same behavior on CIO), server: netty, enabled calls logging
Http client:
```kotlin
fun httpClient() =
    HttpClient(Apache) {
        install(JsonFeature) {
            serializer = JacksonSerializer()
        }

        install(ResponseObserver) {
            onResponse {
                httpClientLogger.trace { "Response received" }
            }
        }

        install(Logging) {
            logger = Logger.TRACE
            level = LogLevel.ALL
        }
    }
```

**Describe the bug**
I have following code:
```kotlin
        runCatching {
            client.get<ConversationInformation> {
                url(endpoint)
                header("Authorization", "Bearer $token")
            }
        }.onFailure {
            logger.error(it) { "It was not possible to fetch conversation information!" }
        }.onSuccess {
            logger.debug { "Successfully got conversation information." }
            logger.trace { it }
        }.getOrNull()
```

And when executed, SOMETIMES it never logs either `onFailure` nor `onSuccess` block.
The client GET request in some cases either silently fails or gets stuck/suspended for indefinite time, not sure how is that possible, but following logs are produced by the call logging mentioned previously.
```
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging - REQUEST: https://<internal URL>
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging - METHOD: HttpMethod(value=GET)
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging - COMMON HEADERS
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging - -> Authorization: Bearer <some bearer>
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging - -> Accept: application/json
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging - -> Accept-Charset: UTF-8
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging - CONTENT HEADERS
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging - BODY Content-Type: null
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging - BODY START
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging -
21/04/2020 12:27:54 [TRACE] com.wire.HttpCallsLogging - BODY END
21/04/2020 12:27:55 [TRACE] com.wire.ObserverLogger - Response received
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - BODY Content-Type: application/json
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - BODY START
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - RESPONSE: 200 OK
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - METHOD: HttpMethod(value=GET)
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - FROM: https://<internal URL>
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - COMMON HEADERS
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - -> Date: Tue, 21 Apr 2020 12:27:54 GMT
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - -> Content-Type: application/json
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - -> Vary: Accept-Encoding
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - -> Content-Length: 125
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - -> Via: 1.1 google
21/04/2020 12:27:55 [TRACE] com.wire.HttpCallsLogging - -> Alt-Svc: clear
```
It seems like the body was never received.

When the  request is successful, log looks like this:
```
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - REQUEST: https://<internal URL>
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - METHOD: HttpMethod(value=GET)
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - COMMON HEADERS
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - -> Authorization: Bearer <some bearer>
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - -> Accept: application/json
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - -> Accept-Charset: UTF-8
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - CONTENT HEADERS
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - BODY Content-Type: null
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - BODY START
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging -
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - BODY END
21/04/2020 12:43:24 [TRACE] com.wire.ObserverLogger - Response received
21/04/2020 12:43:24 [TRACE] com.wire.ObserverLogger - Sending to registry
21/04/2020 12:43:24 [TRACE] com.wire.ObserverLogger - Registered
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - BODY Content-Type: application/json
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - BODY START
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - RESPONSE: 200 OK
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - METHOD: HttpMethod(value=GET)
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - FROM: https://<internal URL>
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - COMMON HEADERS
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - -> Date: Tue, 21 Apr 2020 12:43:24 GMT
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - -> Content-Type: application/json
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - -> Vary: Accept-Encoding
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - -> Content-Length: 125
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - -> Via: 1.1 google
21/04/2020 12:43:24 [TRACE] com.wire.HttpCallsLogging - -> Alt-Svc: clear
21/04/2020 12:43:24 [DEBUG] c.w.b.p.services.ConversationService - Successfully got conversation information.
21/04/2020 12:43:24 [TRACE] c.w.b.p.services.ConversationService - ConversationInformationd(dto=SomeDataHere)
21/04/2020 12:43:25 [TRACE] com.wire.HttpCallsLogging - {"id":"3f7fad1d-5066-487d-8836-9e40928bb12f","name":"a","members":[{"id":"420e4b22-fac2-47e8-b07f-63958d7fcccd","status":0}]}
21/04/2020 12:43:25 [TRACE] com.wire.HttpCallsLogging - BODY END
```

**To Reproduce**
It is happening for like 1/5 of all requests, whole service code available [here](https://github.com/wireapp/poll-bot/tree/staging)

**Expected behavior**
IF the request is not successful, the client should throw exception or eventually time out and then throw exception and not to silently die.  Not sure what is happening though

EDIT: 
tried with 
```kotlin
            val response = client.get<HttpStatement> {
                url(endpoint)
                header("Authorization", "Bearer $token")
            }.execute()
            logger.trace { "Executed" }
```
as well and if it gets stuck, `Executed` is never printed. Also, seems similar to #679 

EDIT2:
I found out that the flow is following:
1) http client executes GET
2) target server respond with 200
3) http client gets stuck
4) target server throws `SocketTimeoutException` after a while

I this case I would expect exception to be thrown.

EDIT3:
this is definitely connected to 
```kotlin
        install(ResponseObserver) {
            onResponse {
                httpClientLogger.trace { "Response received" }
            }
        }
```
as without observer, everything is nice and shiny

EDIT4:
Happens with Jetty and OkHttp engines as well. 
Mostly it happens after multiple requests, like first one or two works, but then it stops working. This is happening in our staging Kubernets.

**Comment by LukasForst** [comment]
Final update:

I tested all engines and it seems that this issue is with all of them. Also, this issue is definitely related to the `ResponseObserver` as without the observer, everything works as expected (60 requests in the loop went fine).

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-410">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.

**Comment by brandonwu** [comment]
For anybody else encountering this:
`ResponseObserver` splits the response body `ByteReadChannel` into two channels so that the hook can read from it without also consuming it from the response returned at the end of the pipeline. But you also need to consume all the bytes in the body channel sent to the hook with
```
it.content.discard()
```
or else bad things will happen (timeouts, blank bodies) because the other channel is stalled.

**Comment by LukasForst** [comment]
Unfortunately I can't confirm that, even in the newest (`1.4.1`) Ktor version. 

Following code results in the same behaviour as previously (silently dying)
```kotlin
        install(ResponseObserver) {
            onResponse {
                httpClientLogger.trace { "Response received" }
                it.content.discard()
            }
        }
```

**Comment by brandonwu** [comment]
Sorry that didn't fix it. I had a related problem and that solved it for me.

The only other thing I would suggest would be _if you don't need access to the body_ in your response hook: implement your own feature without using `ResponseObserver` to install it, avoiding its special behavior (it spawns a coroutine to run your hook).

bare-bones example:
```kotlin
import io.ktor.client.*
import io.ktor.client.features.*
import io.ktor.client.statement.*
import io.ktor.util.*

class PlainLogger(private val loggerPrefix: String) {

    fun log(response: HttpResponse) {
        println("PlainLogger[$loggerPrefix]: ${response.status}")
    }

    class Config {
        var loggerPrefix: String = "PlainLogger"
    }

    companion object Feature : HttpClientFeature<Config, PlainLogger> {

        override val key: AttributeKey<PlainLogger> = AttributeKey("PlainLogger")

        override fun prepare(block: Config.() -> Unit): PlainLogger {
            val config = Config().apply(block)
            return PlainLogger(config.loggerPrefix)
        }

        override fun install(feature: PlainLogger, scope: HttpClient) {
            // Synchronous response pipeline hook, instead of ResponseObserver (which spawns a new coroutine)
            scope.receivePipeline.intercept(HttpReceivePipeline.After) { response ->
                /**
                 * WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
                 * Do not consume [HttpResponse.content] here, or you will corrupt the client response.
                 */
                try {
                    feature.log(response)
                } catch (error: Throwable) {
                    // don't let logger failure fail request
                }
            }
        }
    }
}
```

usage:
```kotlin
import io.ktor.client.*
import io.ktor.client.engine.apache.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import kotlinx.coroutines.runBlocking

val httpClient: HttpClient = HttpClient(Apache) {
    install(PlainLogger) {
        loggerPrefix = "Logger#1"
    }
}

fun main() {
    runBlocking {
        httpClient.get<HttpStatement> {
            url("http://example.com/")
        }.execute()
    }
}
```

output:
```
PlainLogger[Logger#1]: 200 OK
```

**Comment by Stexxe** [comment]
Cannot reproduce it using Ktor 1.6.1.

**Comment by dshatz** [comment]
I am encountering the same issue on 3.1.2 but not on 3.3.1.

Engine - any.
Request - any.
Kotlin - 2.2.21

The `Logging` plugin is causing this. (Removing fixes the issue).

Minimal example:
```kotlin
val ktorClient = HttpClient(OkHttp) {
        expectSuccess = true
        install(ContentNegotiation) {
            json(json)
        }
        install(Logging) {
            logger = Logger.DEFAULT
            level = LogLevel.BODY
        }
        defaultRequest {
            contentType(ContentType.Application.Json)
        }
}
```

**Comment by cs-nuuk** [comment]
I get this issue with 3.3.3 and OkHttp when I try to get HTML. When I try to get JSON using the ContentNegotiation plugin, it does not get stuck. Removing logging helped. Thank you for that.

```
val html: String = client.get(DetailsQuery(id = packageName)).body()
return html
```

**Comment by Stexxe** [comment]
@cs-nuuk, can you please share a complete code snippet to reproduce the problem?

**Comment by cs-nuuk** [comment]
@Stexxe 

This works:
```
@Singleton
class KtorAppDetails @Inject constructor() : AppDetailsDataSource() {
    private val okHttpClient = OkHttpClient.Builder().build()

    private val client = HttpClient(OkHttp) {
        engine {
            preconfigured = okHttpClient
        }

        install(Resources)

        defaultRequest {
            url {
                host = "play.google.com"
                path("/store/apps/")
                protocol = URLProtocol.HTTPS
            }
        }
    }

    override suspend fun getAppDetailsHtml(packageName: String): String {
        val html: String = client.get(DetailsQuery(id = packageName)).body()
        return html
    }
}

```

When I add this, it gets stuck:
```
install(Logging) {
    logger = Logger.ANDROID
    format = LoggingFormat.OkHttp
    level = if (BuildConfig.DEBUG) LogLevel.ALL else LogLevel.NONE
}
```

This is the resource:
```
@Resource("/details")
class DetailsQuery(val id: String)
```

I hope this helps. 🙂

**Comment by cs-nuuk** [comment]
In other data sources where I retrieve JSON and no HTML, the Logger plugin works.

The setup there is very similar. I just add this:

```
install(ContentNegotiation) {
    json(json)
}
```

And then do the requests like this:
```
override suspend fun getTickets(): List<Ticket> {
    val tickets: List<Ticket> = client.get(TicketsQuery.All()).body()
    return tickets
}
```

**Comment by Stexxe** [comment]
This problem ([KTOR-9148](https://youtrack.jetbrains.com/issue/KTOR-9148)) has been fixed in Ktor 3.4.0.
